File: dart/behavioral/memento/guru_example/bin/originator.dart
import 'dart:math';

import 'concrete_memento.dart';
import 'memento.dart';

class Originator {
  String _state;

  Originator(String s) {
    _state = s;
    print('Originator constructor state = $_state');
  }

  void doSomething() {
    _state = generateRandomString(30);
    print('Originator doSomething state = ${_state}');
  }

  String generateRandomString([int length = 10]) {
    const charSet = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';

    Random rand = Random();
    var arr = List(length);
    for (var i = 0; i < arr.length; i++) {
      arr[i] = charSet[rand.nextInt(charSet.length)];
    }
    return arr.join('');
  }

  Memento save() {
    return ConcreteMemento(_state);
  }

  void restore(Memento memento) {
    _state = memento.getState();
    print('Originator restore state = $_state');
  }
}

File: typescript/behavioral/memento/guru_example/originator.ts
import {ConcreteMemento} from "./concreteMemento";
import {Memento} from "./memento";

export class Originator {
  private state: string

  constructor(state: string) {
    this.state = state
    console.log(`Originator state = ${this.state}`)
  }

  public doSomething(): void {
    console.log('Originator doSomething')
    this.generateRandomString(30)
  }

  public generateRandomString(length: number = 10): string {
    const charSet = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
    return Array
      .apply(null, {length})
      .map(() => charSet.charAt(Math.floor(Math.random() * charSet.length)))
      .join('')
  }

  public save(): Memento {
    return new ConcreteMemento(this.state)
  }

  public restore(memento: Memento): void {
    this.state = memento.getState()
    console.log(`Originator state = ${this.state}`)
  }
}
